src/lib.rs

```diff
@@ -41,1346 +53,988 @@ use std::{
+                has_hiding = true;
             }
-        }
 
-        let pc_result = Self::batch_check_individual_opening_challenges(
-            vk,
-            commitments,
-            &poly_query_set,
-            &poly_evals,
-            proof,
-            fs_rng,
-        )?;
+            // TODO: look for arithmetic operations optimization and removing clone
+            // add segment linear combination to overall combination,
+            // both for witnesses, commitments (and their randomnesses, if hiding)
+            combined_polynomial += (cur_challenge, polynomial);
+            if has_hiding {
+                combined_randomness += &(randomness.clone() * &cur_challenge);
+            }
 
-        if !pc_result {
-            eprintln!("Evaluation proofs failed to verify");
-            return Ok(false);
+            // next power of lambda
+            cur_challenge = cur_challenge * &lambda;
         }
 
-        Ok(true)
-    }
-}
+        end_timer!(combine_time);
 
-/// Evaluate the given polynomials at `query_set`.
-pub fn evaluate_query_set<'a, F: Field>(
-    polys: impl IntoIterator<Item = &'a LabeledPolynomial<F>>,
-    query_set: &QuerySet<'a, F>,
-) -> Evaluations<'a, F> {
-    let polys = BTreeMap::from_iter(polys.into_iter().map(|p| (p.label(), p)));
-    let mut evaluations = Evaluations::new();
-    for (label, (_, point)) in query_set {
-        let poly = polys
-            .get(label)
-            .expect("polynomial in evaluated lc is not found");
-        let eval = poly.evaluate(*point);
-        evaluations.insert((label.clone(), *point), eval);
+        Ok((combined_polynomial, combined_randomness, has_hiding))
     }
-    evaluations
-}
 
-/// Evaluate the given polynomials at `query_set` and returns a Vec<((poly_label, point_label), eval)>)
-pub fn evaluate_query_set_to_vec<'a, F: Field>(
-    polys: impl IntoIterator<Item = &'a LabeledPolynomial<F>>,
-    query_set: &QuerySet<'a, F>,
-) -> Vec<((String, String), F)> {
-    // use std::{
-    //     collections::BTreeMap,
-    //     iter::FromIterator,
-    // };
-    let polys = BTreeMap::from_iter(polys.into_iter().map(|p| (p.label(), p)));
-    let mut v = Vec::new();
-    for (label, (point_label, point)) in query_set {
-        let poly = polys
-            .get(label)
-            .expect("polynomial in evaluated lc is not found");
-        let eval = poly.evaluate(*point);
-        v.push(((label.clone(), point_label.clone()), eval));
-    }
-    v
-}
+    /// Build the multi-point multi-poly opening combination.
+    ///       LC_x(p_i(X),h(X)) = Sum_i  lambda^i z_i(x)/z(x) * p_i(X)  - h(X)
+    // TODO: let us declare this function as internal.
+    fn multi_point_multi_poly_open_combine<'a>(
+        ck: &Self::CommitterKey,
+        labeled_polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField>>,
+        labeled_commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
```

Comment:
As pointed out in my comment below: why do we pass the commitments here?

Reply:
Removed